### PR TITLE
Tickets/dm 37262

### PR DESCRIPTION
**Pull request by teweitsai**
* Adapt the **ts_m2com** v0.8.1 to use the commands related to ILC, CLC, etc. directly.
* Remove the dependency of **ts_salobj**.
* Add the **transition_local_mode.uml**.

### CODE REVIEW

File: python/lsst/ts/m2gui/model.py [hunk 1]
```diff
@@ -1077,6 +1048,127 @@ def _process_lost_connection(self):
...
+
+        await self.controller.set_closed_loop_control_mode(ClosedLoopControlMode.Idle)
+        await self.controller.power(PowerType.Communication, True)
+
+        self.local_mode = LocalMode.Diagnostic
+
+        self.report_control_status()
+
+    async def enable(self):
+        """Transition from the Diagnostic mode to the Enable mode.
```
Comment (pkubanek) [inline python/lsst/ts/m2gui/model.py:1073]: I am not sure enabled/disabled states should be misused for diagnostic mode. M1M3 has (I believe) similar problem, yet please look https://ts-m1m3support.lsst.io/dir_633db867517fc930a7645f2dac5f41c1.html how diagnostic (aka Engineering) mode is implemented.

I believe those are hooks for SAL state transition commands the class is acting on. Then disable is meant for the mirror state when it cannot be commanded. Enabled is for states when the mirror can be commanded. The granularity of accepted commands is then ruled by sub-state, which can be parked, EngineeringParked (diagnostic parked), ...

**Review by pkubanek — CHANGES_REQUESTED** [review]
Please either explain why disable/enable commands can be used to enter/exit diagnostic mode, or follow the M1M3 model on EngineeringMode,..

File: doc/developer-guide/developer-guide.rst [hunk 2]
```diff
@@ -11,12 +11,22 @@ This GUI is constructed on the top of Qt framework (`Qt for Python <https://wiki
...
-* `ts_utils <https://github.com/lsst-ts/ts_utils>`_
-* `ts_config_mttcs <https://github.com/lsst-ts/ts_config_mttcs>`_
 * `ts_m2com <https://github.com/lsst-ts/ts_m2com>`_
 
+.. _State:
+
+State
+=====
+
+The state machine is list below.
```
[the remark this answers is not in the document]
Comment (teweitsai) [reply doc/developer-guide/developer-guide.rst]: Fixed in the commit: [a3d07b5](https://github.com/lsst-ts/ts_m2gui/pull/28/commits/a3d07b5f5bcc6f0a78bf6dc27e265bdbbe5f4714). Thanks!

File: python/lsst/ts/m2gui/model.py (hunk 1 above)
Comment (teweitsai) [reply python/lsst/ts/m2gui/model.py:1073]: As the discussion, renamed the state transition functions in commit: [b10208d](https://github.com/lsst-ts/ts_m2gui/pull/28/commits/b10208d77d2457fe1d385c7f87f4f7a64ec7a531). Thanks!

File: doc/uml/class_m2gui.uml [hunk 3]
```diff
@@ -26,6 +26,7 @@ MainWindow ..> QMessageBoxAsync
 Model *-- FaultManager
 Model *-- UtilityMonitor
 Model *-- SignalControl : emit()
+Model *-- SignalPowerSystem : emit()
```
Comment (pkubanek) [inline doc/uml/class_m2gui.uml:29]: This is misleading. You aren't emitting a class, you are emitting its member, is_state_updated. But also that design makes your code much harder to read and understand. Let's discuss that in person once I am back.

File: python/lsst/ts/m2gui/controltab/tab_utility_view.py [hunk 4]
```diff
@@ -371,6 +384,58 @@ def _create_group_displacements(self):
...
+        ----------
+        signal_power_system : `SignalPowerSystem`
+            Signal of the power system.
+        """
+        signal_power_system.is_state_updated.connect(
+            self._callback_update_power_system_status
+        )
+
+    @asyncSlot()
+    async def _callback_update_power_system_status(self, is_state_updated):
```
Comment (pkubanek) [inline python/lsst/ts/m2gui/controltab/tab_utility_view.py:428]: This round of private functions can be simplified by passing data with signal. More in person

File: tests/layout/test_layout_control_mode.py [hunk 5]
```diff
@@ -93,16 +92,11 @@ async def test_callback_signal_control_prohibit_control(qtbot, widget):
...
-    )
-    await controller.write_command_to_server(
-        "enable", controller_state_expected=salobj.State.ENABLED
-    )
+    await widget_async.model.enter_diagnostic()
+    await widget_async.model.enter_enable()
 
     # Sleep so the event loop can access CPU to handle the signal
-    await asyncio.sleep(1)
+    await asyncio.sleep(15)
```
Comment (pkubanek) [inline tests/layout/test_layout_control_mode.py:99]: Does it really need to be 15 seconds? enter_enable should be much faster, shouldn't it?

**Review by pkubanek — APPROVED** [review]
Looks good, but please see comments.

File: doc/uml/class_m2gui.uml (hunk 3 above)
Comment (teweitsai) [reply doc/uml/class_m2gui.uml:29]: I think this is what I could do for the class UML to explain the relationship among classes/signals. The documents and class diagrams are provided to let the developers know the code architecture. Thanks!

File: python/lsst/ts/m2gui/controltab/tab_utility_view.py (hunk 4 above)
Comment (teweitsai) [reply python/lsst/ts/m2gui/controltab/tab_utility_view.py:428]: This is to be consistent in the code base to improve the code maintainability. Thanks!

File: tests/layout/test_layout_control_mode.py (hunk 5 above)
Comment (teweitsai) [reply tests/layout/test_layout_control_mode.py:99]: This is for the event loop to handle the incoming telemetry and events, which takes some time. Thanks!

File: python/lsst/ts/m2gui/model.py [hunk 6]
```diff
@@ -1077,6 +1048,127 @@ def _process_lost_connection(self):
...
+        `RuntimeError`
+            When the system is not in the Enable mode.
+        """
+
+        if self.local_mode != LocalMode.Enable:
+            raise RuntimeError(
+                f"System is in {self.local_mode!r} instead of {LocalMode.Enable!r}."
+            )
+
+        await self._basic_cleanup_and_power_off_motor()
```
[the remark this answers is not in the document]
Comment (teweitsai) [reply python/lsst/ts/m2gui/model.py:1113]: It is possible to combine to a single command. However, I felt two separate commands might be more clear. Thanks!

**Pull request merged**